ecc/bls12381/g1.go

```diff
@@ -177,7 +185,35 @@ func (g *G1) ScalarMult(k *Scalar, P *G1) { g.scalarMult(k.Bytes(), P) }
 func (g *G1) scalarMult(k []byte, P *G1) {
 	var Q G1
 	Q.SetIdentity()
-	for i := 8*len(k) - 1; i >= 0; i-- {
+	T := &G1{}
+	var mults [16]G1
+	mults[0].SetIdentity()
+	mults[1].Set(P)
+	for i := 1; i < 8; i++ {
+		mults[2*i].Set(&mults[i])
+		mults[2*i].Double()
+		mults[2*i+1].Add(&mults[2*i], P)
+	}
+	for i := 8*len(k) - 4; i >= 0; i -= 4 {
+		Q.Double()
+		Q.Double()
+		Q.Double()
+		Q.Double()
+		idx := 0xf & (k[i/8] >> uint(i%8))
+		for j := 0; j < 16; j++ {
+			T.cmov(&mults[j], subtle.ConstantTimeByteEq(idx, uint8(j)))
+		}
+		Q.Add(&Q, T)
+	}
+	g.Set(&Q)
+}
+
+// scalarMultShort multiplies by short, constant scalars.
+func (g *G1) scalarMultShort(k []byte, P *G1) {
+	// Since the scalar is short and low Hamming weight not much helps.
```

Comment:
Addition chains might help — in this case I found one with one double less, which is not worth the trouble.